aptos-move/framework/aptos-framework/sources/fungible_asset.move

```diff
@@ -656,6 +730,30 @@ module aptos_framework::fungible_asset {
         move_to(&metadata_object_signer, supply);
     }
 
+    public fun upgrade_store_to_concurrent<T: key>(
+        owner: &signer,
+        store: Object<T>,
+    ) acquires FungibleStore {
+        assert!(object::owns(store, signer::address_of(owner)), error::permission_denied(ENOT_STORE_OWNER));
+        assert!(!is_frozen(store), error::invalid_argument(ESTORE_IS_FROZEN));
+        assert!(features::concurrent_fungible_assets_enabled(), error::invalid_argument(ECONCURRENT_BALANCE_NOT_ENABLED));
+        ensure_store_upgraded_to_concurrent_internal(object::object_address(&store));
+    }
+
+    /// Ensure a known `FungibleStore` has `ConcurrentFungibleBalance`.
+    inline fun ensure_store_upgraded_to_concurrent_internal(
+        fungible_store_address: address,
+    ) acquires FungibleStore {
+        if (!exists<ConcurrentFungibleBalance>(fungible_store_address)) {
```

Comment:
why the nest... can't we just exit or assert this hasn't been called... so said to see this type of nesting... I have to read the end of a massive blob to see there's no else.

Reply:
@davidiw thanks for flagging. The below suggestion eliminates a bytecode instruction:

```rust
    /// Ensure a known `FungibleStore` has `ConcurrentFungibleBalance`.
    inline fun ensure_store_upgraded_to_concurrent_internal(
        fungible_store_address: address,
    ) acquires FungibleStore {
        if (exists<ConcurrentFungibleBalance>(fungible_store_address)) {
            // Do nothing, return statements not currently supported for inline.
        } else {
            let store = borrow_global_mut<FungibleStore>(fungible_store_address);
            let balance = aggregator_v2::create_unbounded_aggregator();
            aggregator_v2::add(&mut balance, store.balance);
            store.balance = 0;
            let object_signer = create_signer::create_signer(fungible_store_address);
            move_to(&object_signer, ConcurrentFungibleBalance { balance });
        };
    }
```

# Current bytecode result from1c38359fa9

Taken from the disassembly of `upgrade_store_to_concurrent()`

```asm
	27: Exists[1](ConcurrentFungibleBalance)
	28: Not ; This is the instruction that gets eliminated
	29: BrFalse(51)
B10: ; This is where `ensure_store_upgraded_to_concurrent_internal()` begins
	30: CopyLoc[3](loc1: address)
	31: MutBorrowGlobal[10](FungibleStore)
	32: StLoc[5](loc3: &mut FungibleStore)
```

# Improved bytecode result from suggestion

Yields

```asm
	27: Exists[1](ConcurrentFungibleBalance)
	28: BrFalse(30)
B10:
	29: Branch(51)
B11: ; This is where `ensure_store_upgraded_to_concurrent_internal()` begins
	30: CopyLoc[3](loc1: address)
	31: MutBorrowGlobal[10](FungibleStore)
	32: StLoc[5](loc3: &mut FungibleStore)
```

# Steps to reproduce

```sh

```

Reply:
ensure_store_upgraded_to_concurrent_internal is not going to be called often, with making it opt-in, so leaving it readable